### PR TITLE
make redis module name and start_link args configurable

### LINKED ISSUE #292: Add support for redis sentinel
We have been using the [fork](https://github.com/akira/exq/compare/master...ananthakumaran:sentinel) at production for sometime. If there is interest, I can prepare a PR, though we need to figure out which approach needs to be taken. [redix_sentinel](https://github.com/ananthakumaran/redix_sentinel) is api compatible with redix except for the start_link method. so we could make the module_name & start_link args configurable. At this point, I am not sure whether sentinel support would be added in redix, see https://github.com/whatyouhide/redix/issues/51. I am open to other ideas also.

### PR DESCRIPTION
**Pull request by ananthakumaran**
fixes #292

### CODE REVIEW

**Comment by zhongwencool** [comment]
`MFA` call make this function be slowly than before. :(
This module can be sure at compile time.
how about:
```elixir
defmodule Exq.Connection  do
{ok, Module} = Application.get_env(:exq, :redis_worker)
def q(redis, command) do
    Module.command(redis, command, [timeout: Config.get(:redis_timeout)])
end		   
end
```

**Comment by ananthakumaran** [comment]
```elixir
defmodule MFABench do
  use Benchfella

  setup_all do
    {:ok, pid} = Exq.start_link
  end

  bench "mfa" do
    {:ok, "PONG"} = Exq.Redis.Connection.q(Exq.Redis.Client, ["PING"])
  end

  bench "direct" do
    {:ok, "PONG"} = Exq.Redis.Connection.qdirect(Exq.Redis.Client, ["PING"])
  end

  teardown_all pid do
    Exq.stop(pid)
  end
end
```
```
Settings:
  duration:      1.0 s

## MFABench
[16:44:47] 1/2: direct
[16:44:49] 2/2: mfa

Finished in 5.49 seconds

## MFABench
benchma iterations   average time
direct       20000   87.52 µs/op
mfa          20000   88.59 µs/op
```

As you can see from a sample run, the difference is negligible. The difference between a single MFA call and normal call would normally become negligible when there is a network call involved.

As for why not figure out the Module during compile time?

Because it does't play nice with many deployment strategies. Most users compile the code in CI server and then different configs would be used for different env. Basically the config would be available just before the app startup, not during the compile step that happens a lot earlier. Compile time config would also make it complicated to build docker images, because you would need to build different docker image for each environment, instead of just changing the config.

File: test/test_helper.exs [hunk 1]
```diff
@@ -96,13 +106,15 @@ defmodule TestRedis do
   def start do
     unless Config.get(:test_with_local_redis) == false do
       [] = :os.cmd('redis-server test/test-redis.conf')
-      :timer.sleep(100)
+      [] = :os.cmd('redis-server test/test-sentinel.conf --sentinel')
```
Comment (akira) [inline test/test_helper.exs]: Do we need to start Redis Sentinel for everything?  Maybe we should only start for those few tests?

**Comment by akira** [comment]
@ananthakumaran thanks for the performance benchmark, given the small difference we can leave it as an MFA call.

File: test/test_helper.exs (hunk 1 above)
Comment (ananthakumaran) [reply test/test_helper.exs]: This is called only once at the beginning before any of the tests starts and stop is invoked at the end when all the tests are executed. Throughput wise delaying the start of sentinel server won't make any difference.

**Comment by akira** [comment]
I seem to be getting some errors locally when running tests:

```
    ** (EXIT from #PID<0.3398.0>) an exception was raised:
         ** (FunctionClauseError) no function clause matching in Redix.Connection.handle_call/3
             lib/redix/connection.ex:183: Redix.Connection.handle_call(:node, {#PID<0.3398.0>, #Reference<0.0.2.8767>}, %Redix.Connection{backoff_current: nil, opts: [socket_opts: [], sync_connect: false, backoff_initial: 500, backoff_max: 30000, log: [disconnection: :error, failed_connection: :error, reconnection: :info], exit_on_disconnection: false, host: "127.0.0.1", port: 6555], receiver: #PID<0.3401.0>, shared_state: #PID<0.3400.0>, socket: #Port<0.13906>})
             lib/connection.ex:488: Connection.handle_call/3
             (stdlib) gen_server.erl:629: :gen_server.try_handle_call/4
             (stdlib) gen_server.erl:661: :gen_server.handle_msg/5
             (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
```

Have you seen this?

File: test/test_helper.exs (hunk 1 above)
Comment (akira) [reply test/test_helper.exs]: Ah ok, my bad, makes sense.

**Comment by ananthakumaran** [comment]
No, does it happen on any specific test? can you provide more details?

**Comment by akira** [comment]
About 90 failures:

```
87) test to_job_serialized using module atom (JobQueueTest)
     test/job_queue_test.exs:158
     ** (EXIT from #PID<0.2404.0>) an exception was raised:
         ** (FunctionClauseError) no function clause matching in Redix.Connection.handle_call/3
             lib/redix/connection.ex:183: Redix.Connection.handle_call(:node, {#PID<0.2404.0>, #Reference<0.0.2.4945>}, %Redix.Connection{backoff_current: nil, opts: [socket_opts: [], sync_connect: false, backoff_initial: 500, backoff_max: 30000, log: [disconnection: :error, failed_connection: :error, reconnection: :info], exit_on_disconnection: false, host: "127.0.0.1", port: 6555], receiver: #PID<0.2407.0>, shared_state: #PID<0.2406.0>, socket: #Port<0.34451>})
             lib/connection.ex:488: Connection.handle_call/3
             (stdlib) gen_server.erl:629: :gen_server.try_handle_call/4
             (stdlib) gen_server.erl:661: :gen_server.handle_msg/5
             (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3

     The following output was logged:

     2018-01-15 21:05:23.076 [error]: GenServer #PID<0.2405.0> terminating
     ** (FunctionClauseError) no function clause matching in Redix.Connection.handle_call/3
         lib/redix/connection.ex:183: Redix.Connection.handle_call(:node, {#PID<0.2404.0>, #Reference<0.0.2.4945>}, %Redix.Connection{backoff_current: nil, opts: [socket_opts: [], sync_connect: false, backoff_initial: 500, backoff_max: 30000, log: [disconnection: :error, failed_connection: :error, reconnection: :info], exit_on_disconnection: false, host: "127.0.0.1", port: 6555], receiver: #PID<0.2407.0>, shared_state: #PID<0.2406.0>, socket: #Port<0.34451>})
         lib/connection.ex:488: Connection.handle_call/3
         (stdlib) gen_server.erl:629: :gen_server.try_handle_call/4
         (stdlib) gen_server.erl:661: :gen_server.handle_msg/5
         (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
     Last message: :node
     State: %Redix.Connection{backoff_current: nil, opts: [socket_opts: [], sync_connect: false, backoff_initial: 500, backoff_max: 30000, log: [disconnection: :error, failed_connection: :error, reconnection: :info], exit_on_disconnection: false, host: "127.0.0.1", port: 6555], receiver: #PID<0.2407.0>, shared_state: #PID<0.2406.0>, socket: #Port<0.34451>}



 88) test to_job_serialized using module string (JobQueueTest)
     test/job_queue_test.exs:165
     ** (EXIT from #PID<0.2411.0>) an exception was raised:
         ** (FunctionClauseError) no function clause matching in Redix.Connection.handle_call/3
             lib/redix/connection.ex:183: Redix.Connection.handle_call(:node, {#PID<0.2411.0>, #Reference<0.0.2.4978>}, %Redix.Connection{backoff_current: nil, opts: [socket_opts: [], sync_connect: false, backoff_initial: 500, backoff_max: 30000, log: [disconnection: :error, failed_connection: :error, reconnection: :info], exit_on_disconnection: false, host: "127.0.0.1", port: 6555], receiver: #PID<0.2414.0>, shared_state: #PID<0.2413.0>, socket: #Port<0.34452>})
             lib/connection.ex:488: Connection.handle_call/3
             (stdlib) gen_server.erl:629: :gen_server.try_handle_call/4
             (stdlib) gen_server.erl:661: :gen_server.handle_msg/5
             (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3

     The following output was logged:

     2018-01-15 21:05:23.078 [error]: GenServer #PID<0.2412.0> terminating
     ** (FunctionClauseError) no function clause matching in Redix.Connection.handle_call/3
         lib/redix/connection.ex:183: Redix.Connection.handle_call(:node, {#PID<0.2411.0>, #Reference<0.0.2.4978>}, %Redix.Connection{backoff_current: nil, opts: [socket_opts: [], sync_connect: false, backoff_initial: 500, backoff_max: 30000, log: [disconnection: :error, failed_connection: :error, reconnection: :info], exit_on_disconnection: false, host: "127.0.0.1", port: 6555], receiver: #PID<0.2414.0>, shared_state: #PID<0.2413.0>, socket: #Port<0.34452>})
         lib/connection.ex:488: Connection.handle_call/3
         (stdlib) gen_server.erl:629: :gen_server.try_handle_call/4
         (stdlib) gen_server.erl:661: :gen_server.handle_msg/5
         (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
     Last message: :node
     State: %Redix.Connection{backoff_current: nil, opts: [socket_opts: [], sync_connect: false, backoff_initial: 500, backoff_max: 30000, log: [disconnection: :error, failed_connection: :error, reconnection: :info], exit_on_disconnection: false, host: "127.0.0.1", port: 6555], receiver: #PID<0.2414.0>, shared_state: #PID<0.2413.0>, socket: #Port<0.34452>}

......

 89) test default redis_worker_opts (Exq.ConfigTest)
     test/config_test.exs:111
     ** (MatchError) no match of right hand side value: {RedixSentinel, [[role: "master", group: "exq", sentinels: [[host: "127.0.0.1", port: 6666]]], [database: 0, password: nil], [backoff: 100, timeout: 5000, name: Exq.Redis.Client, socket_opts: []]], [scheduler_enable: true, namespace: "exq", scheduler_poll_timeout: 200, workers_sup: Exq.Worker.Sup, poll_timeout: 100, enqueuer: Exq.Enqueuer, metadata: Exq.Worker.Metadata, stats: Exq.Stats, name: nil, scheduler: Exq.Scheduler, queues: ["default"], redis: Exq.Redis.Client, concurrency: [{"default", 100, 0}], middleware: Exq.Middleware.Server, default_middleware: [Exq.Middleware.Stats, Exq.Middleware.Job, Exq.Middleware.Manager], mode: :default, shutdown_timeout: 7000]}
     stacktrace:
       test/config_test.exs:124: (test)



 90) test api redis_worker_opts (Exq.ConfigTest)
     test/config_test.exs:154
     ** (MatchError) no match of right hand side value: {RedixSentinel, [[role: "master", group: "exq", sentinels: [[host: "127.0.0.1", port: 6666]]], [database: 0, password: nil], [backoff: 100, timeout: 5000, name: Exq.Redis.Client, socket_opts: []]], [name: nil, namespace: "test", redis: Exq.Redis.Client, mode: :api]}
     stacktrace:
       test/config_test.exs:156: (test)

.

Finished in 2.7 seconds
144 tests, 90 failures, 6 skipped
```

What version of Redis does this require?  Here is my version:

```
redis-server --version
Redis server v=2.8.3 sha=00000000:0 malloc=libc bits=64 build=e836d8ad888e21a1
```

**Comment by ananthakumaran** [comment]
2.8 does support sentinel. This seems more like a compile time issue? can you delete the _build directory and try to compile again? I suspect something is going wrong with the `Connection.call(conn, :node)` that happens in RedixSentinel module, somehow ends up calling Redix.Connection.call?

**Comment by akira** [comment]
Sure, I just removed `_build` directory and tried again, but it still happens...   Yeah, I noticed this works in Travis. What version of Elixir are you using BTW?

**Comment by ananthakumaran** [comment]
```
~|⇒ elixir --version
Erlang/OTP 20 [erts-9.1.5] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]

Elixir 1.6.0-dev (7e161b5) (compiled with OTP 20)
~|⇒ redis-server --version
Redis server v=4.0.2 sha=00000000:0 malloc=libc bits=64 build=7db4e866648bf9c0
```

**Comment by ananthakumaran** [comment]
I switched to 1.5.1, the test still passes for me

```
~/work/repos/exq|[git:start_link_configurable] ⇒ mix test --no-start
warning: variable "deps" does not exist and is being expanded to "deps()", please use parentheses to remove the ambiguity or change the variable name
  /Users/ananthakumaran/work/repos/exq/deps/flaky_connection/mix.exs:10

===> Compiling idna
===> Compiling mimerl
==> connection
Compiling 1 file (.ex)
Generated connection app
==> redix
Compiling 8 files (.ex)
Generated redix app
===> Compiling metrics
==> exq
 GEN    clean
 GEN    clean-app
 GEN    clean-ct
 ERLC   ranch.erl ranch_acceptor.erl ranch_acceptors_sup.erl ranch_app.erl ranch_conns_sup.erl ranch_listener_sup.erl ranch_protocol.erl ranch_server.erl ranch_ssl.erl ranch_sup.erl ranch_tcp.erl ranch_transport.erl
 APP    ranch.app.src
==> flaky_connection
Compiling 2 files (.ex)
warning: variable "make_ref" does not exist and is being expanded to "make_ref()", please use parentheses to remove the ambiguity or change the variable name
  lib/flaky_connection.ex:6

warning: variable "self" does not exist and is being expanded to "self()", please use parentheses to remove the ambiguity or change the variable name
  lib/flaky_connection_handler.ex:13

warning: variable "self" does not exist and is being expanded to "self()", please use parentheses to remove the ambiguity or change the variable name
  lib/flaky_connection_handler.ex:16

warning: Dict.put/3 is deprecated, use the Map module for working with maps or the Keyword module for working with keyword lists
  lib/flaky_connection_handler.ex:21

warning: variable "self" does not exist and is being expanded to "self()", please use parentheses to remove the ambiguity or change the variable name
  lib/flaky_connection_handler.ex:44

Generated flaky_connection app
warning: String.strip/1 is deprecated, use String.trim/1
  /Users/ananthakumaran/work/repos/exq/deps/poison/mix.exs:4

warning: variable "deps" does not exist and is being expanded to "deps()", please use parentheses to remove the ambiguity or change the variable name
  /Users/ananthakumaran/work/repos/exq/deps/poison/mix.exs:11

warning: variable "package" does not exist and is being expanded to "package()", please use parentheses to remove the ambiguity or change the variable name
  /Users/ananthakumaran/work/repos/exq/deps/poison/mix.exs:12

==> poison
Compiling 4 files (.ex)
warning: Integer.to_char_list/2 is deprecated, use Integer.to_charlist/2
  lib/poison/encoder.ex:161

warning: HashDict.size/1 is deprecated, use maps and the Map module instead
  lib/poison/encoder.ex:283

Generated poison app
==> ssl_verify_fun (compile)
Compiled src/ssl_verify_util.erl
Compiled src/ssl_verify_fingerprint.erl
Compiled src/ssl_verify_pk.erl
Compiled src/ssl_verify_hostname.erl
==> jsx
Compiling 9 files (.erl)
Generated jsx app
==> exjsx
Compiling 1 file (.ex)
warning: HashDict.to_list/1 is deprecated, use maps and the Map module instead
  lib/jsx.ex:143

Generated exjsx app
===> Compiling certifi
===> Compiling hackney
==> excoveralls
Compiling 22 files (.ex)
warning: String.to_char_list/1 is deprecated, use String.to_charlist/1
  lib/excoveralls/cover.ex:12

warning: unexpected beginning of EEx tag "<%=" on "<%= 0 == line.coverage -> %>", please remove "=" accordingly
  lib/templates/html/htmlcov/coverage.html.eex:59

warning: unexpected beginning of EEx tag "<%=" on "<%= true -> %>", please remove "=" accordingly
  lib/templates/html/htmlcov/coverage.html.eex:65

warning: String.to_char_list/1 is deprecated, use String.to_charlist/1
  lib/excoveralls/stats.ex:106

Generated excoveralls app
==> redix_sentinel
Compiling 2 files (.ex)
warning: Enum.filter_map/3 is deprecated, use Enum.filter/2 + Enum.map/2 or for comprehensions
  lib/redix_sentinel.ex:311

Generated redix_sentinel app
==> uuid
Compiling 1 file (.ex)
Generated uuid app
==> exq
Compiling 39 files (.ex)
Generated exq app
Excluding tags: [failure_scenarios: true, pending: true]

warning: variable "server_opts" is unused
  test/config_test.exs:166

warning: this expression will fail with ArithmeticError
  test/performance_test.exs:30

warning: no clause will ever match
  test/performance_test.exs:31
warning: this expression will fail with ArithmeticError
  test/worker_test.exs:49

warning: no clause will ever match
  test/worker_test.exs:55

..........................................................................................................................................

Finished in 7.6 seconds
144 tests, 0 failures, 6 skipped

Randomized with seed 0
mix test --no-start  10.55s user 5.19s system 57% cpu 27.550 total
~/work/repos/exq|[git:start_link_configurable] ⇒ elixir --version
Erlang/OTP 20 [erts-9.1.5] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]

Elixir 1.5.1
~/work/repos/exq|[git:start_link_configurable] ⇒

**Comment by akira** [comment]
OK, seemed to work after Redis upgrade. 

LGTM  👍 👏

**Comment by akira** [comment]
@ananthakumaran would be useful to add to README explaining how to setup Sentinel

**Comment by ananthakumaran** [comment]
sure, I will open a PR for readme change

**Pull request merged**